### PR TITLE
AP_Scripting: system for switching to alternative versions of lua scripts

### PR DESCRIPTION
**Pull request by tridge**
working on this with @Hwurzburg

### CODE REVIEW

**Comment by timtuxworth** [comment]
This seems like a great idea in principle, but it seems like this implementation will delete all other scripts that might have been set up in the scripts directory when the switch is flipped, including applets or any other scripts that might have nothing to do what what is being "switched". Am I wrong about that? If so, perhaps this could have unfortunate unintended consequences.

Of course you could say that all scripts you want to keep all the time must have copies in all the relevant sub-directories. That really doesn't sound like a manageable idea, it seems ripe for accidentally going wrong.

File: libraries/AP_Scripting/lua_bindings.h [hunk 1]
```diff
@@ -10,3 +10,5 @@ int lua_get_i2c_device(lua_State *L);
 int AP_HAL__I2CDevice_read_registers(lua_State *L);
 int lua_get_CAN_device(lua_State *L);
 int lua_get_CAN_device2(lua_State *L);
+int lua_dirlist(lua_State *L);
+int lua_removefile(lua_State *L);
```
Comment (IamPete1) [inline libraries/AP_Scripting/lua_bindings.h:14]: There is a `os.remove()` in core lua we could enable.

Comment (IamPete1) [reply libraries/AP_Scripting/lua_bindings.h:14]: or at least you could bind this function to that.

**Comment by Hwurzburg** [comment]
Finished it up and added the scripting restart binding....tweaked the user messages and removed the debug stuff....tested a bunch in SITL....working fine now...have to add docs...

**Comment by Hwurzburg** [comment]
@trdige one fly in the ointment....while runtime switching 100% in SITL, it does not fully on a real flight controller....the reason is parameter creation.....they have to be done apparently from a reboot....so while things are switched, and scripting restarted, the new params are not setup until a reboot happens....not a biggie in my opinion....still allows switching scripts depending on the mission, just not without a reboot...unless new params are not involved....

**Comment by Hwurzburg** [comment]
flight test https://youtu.be/Zhq7FZWd8c8

**Review by Hwurzburg — COMMENTED** [review]
ready to merge

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 2]
```diff
@@ -356,6 +356,9 @@ userdata Parameter method set_and_save boolean float'skip_check
 userdata Parameter method configured boolean
 userdata Parameter method set_default boolean float'skip_check
 
+include AP_Scripting/AP_Scripting.h
+singleton AP_Scripting rename scripting
+singleton AP_Scripting method scripting_restart void
```
Comment (IamPete1) [inline libraries/AP_Scripting/generator/description/bindings.desc:361]: Maybe just `restart`? So `scripting:restart()`

Comment (Hwurzburg) [reply libraries/AP_Scripting/generator/description/bindings.desc]: done

**Comment by Hwurzburg** [comment]
@tridge I have coded a standalone  param selector/loade applet script that could be integrated so not only are the subdir's scripts swaped in, but a params.param file in the that directory read and set/saved...  will do that as a future enhancement

File: libraries/AP_Scripting/applets/Script_Controller.lua [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+function remove_scripts(subdir)
+   local sdir = get_scripts_dir()
+   local dlist = dirlist(sdir)
+   if not dlist then
+      return false
+   end
+   local ret = false
+   for k,v in ipairs(dlist) do
+      local suffix = v:sub(-4)
+      if suffix == ".lua" and v ~= THIS_SCRIPT then
```
Comment (tridge) [inline libraries/AP_Scripting/applets/Script_Controller.lua:108]: we need to make these comparisons case-insensitive with the upper() method

File: libraries/AP_Scripting/applets/Script_Controller.lua [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+function copy_scripts(subdir)
+   local dlist = dirlist(subdir)
+   if not dlist then
+      return false
+   end
+   local ret = false
+   local sdir = get_scripts_dir()
+   for k, v in ipairs(dlist) do
+      local suffix = v:sub(-4)
+      if suffix == ".lua" and v ~= THIS_SCRIPT then
```
Comment (tridge) [inline libraries/AP_Scripting/applets/Script_Controller.lua:131]: needs to be case insensitive

File: libraries/AP_Scripting/applets/Script_Controller.lua [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+   -- step1, remove lua files from scripts/ that are not in the given subdirectory
+   local subdir = get_scripts_dir() .. "/" .. n
+   local changes_made = remove_scripts(subdir)
+   changes_made = changes_made or copy_scripts(subdir)  --copy files if different
+   return changes_made
+end
+
+local sw_last = 0
+function update()
+   local sw_current = scripting_dir_select:norm_input_ignore_trim() + 2
```
Comment (tridge) [inline libraries/AP_Scripting/applets/Script_Controller.lua:159]: this needs to use the get_aux_cached() method
see aux_cached.lua example

File: libraries/AP_Scripting/applets/Script_Controller.lua (hunk 3 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/Script_Controller.lua]: done

File: libraries/AP_Scripting/applets/Script_Controller.lua (hunk 4 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/Script_Controller.lua]: done

File: libraries/AP_Scripting/applets/Script_Controller.lua (hunk 5 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/Script_Controller.lua]: done

File: libraries/AP_Scripting/lua_bindings.h (hunk 1 above)
Comment (tridge) [reply libraries/AP_Scripting/lua_bindings.h:14]: we will try to enable os, with just remove and rename

File: libraries/AP_Scripting/applets/Script_Controller.lua [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+      return false
+   end
+   f:close()
+   return true
+end
+
+--[[ compare strings case insensitive and return true if match
+
+--]]
+function compare(a,b)
```
Comment (peterbarker) [inline libraries/AP_Scripting/applets/Script_Controller.lua:98]:
```suggestion
function directory_strings_equal(a,b)
```

or something... anyway, something more than just "compare", which tells you nothing about what it returns.  Remembering that c-style comparisons return -1, 0 and 1!

File: libraries/AP_Scripting/applets/Script_Controller.lua [hunk 7]
```diff
@@ -0,0 +1,195 @@
+--[[
+   a script to select other lua scripts using an auxillary switch
+--]]
+
+local THIS_SCRIPT = "Script_Controller.lua"
+local SEL_CH = 302
+local scripting_dir_select = assert(rc:find_channel_for_option(SEL_CH), 'no sw assigned to 302 for script directory selection') -- directory selection channel
```
Comment (tridge) [inline libraries/AP_Scripting/applets/Script_Controller.lua:7]: we don't need this with the aux cached method

File: libraries/AP_Scripting/applets/Script_Controller.lua (hunk 6 above)
Comment (tridge) [inline libraries/AP_Scripting/applets/Script_Controller.lua:98]: better to name this compare_string_ci()

File: libraries/AP_Scripting/applets/Script_Controller.lua [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+   end
+   f:close()
+   return true
+end
+
+--[[ compare strings case insensitive and return true if match
+
+--]]
+function compare(a,b)
+   if string.upper(a) == string.upper(b) then
```
Comment (tridge) [inline libraries/AP_Scripting/applets/Script_Controller.lua:99]: can just do "return string.upper(a) == string.upper(b)"

File: libraries/AP_Scripting/AP_Scripting.cpp [hunk 9]
```diff
@@ -317,6 +317,12 @@ bool AP_Scripting::arming_checks(size_t buflen, char *buffer) const
     return true;
 }
 
+void AP_Scripting::restart()
+{
+    _stop = true;
+    _restart = true;
```
Comment (tridge) [inline libraries/AP_Scripting/AP_Scripting.cpp:323]: possible race? may need to set _restart first

File: libraries/AP_Scripting/applets/Script_Controller.lua (hunk 7 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/Script_Controller.lua:7]: done

File: libraries/AP_Scripting/applets/Script_Controller.lua (hunk 6 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/Script_Controller.lua:98]: done

File: libraries/AP_Scripting/applets/Script_Controller.lua (hunk 8 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/Script_Controller.lua:99]: done

File: libraries/AP_Scripting/AP_Scripting.cpp (hunk 9 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/AP_Scripting.cpp:323]: done

File: libraries/AP_Scripting/lua_bindings.h (hunk 1 above)
Comment (tridge) [reply libraries/AP_Scripting/lua_bindings.h]: done

**Comment by tridge** [comment]
@Hwurzburg I've added os.remove() and os.rename(), seems to work

File: libraries/AP_Scripting/AP_Scripting.cpp [hunk 10]
```diff
@@ -317,6 +317,12 @@ bool AP_Scripting::arming_checks(size_t buflen, char *buffer) const
     return true;
 }
 
+void AP_Scripting::restart()
```
Comment (tridge) [inline libraries/AP_Scripting/AP_Scripting.cpp:320]: rename to restart_all() (sugestion from @IamPete1 )

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 11]
```diff
@@ -356,6 +356,9 @@ userdata Parameter method set_and_save boolean float'skip_check
 userdata Parameter method configured boolean
 userdata Parameter method set_default boolean float'skip_check
 
+include AP_Scripting/AP_Scripting.h
+singleton AP_Scripting rename scripting
+singleton AP_Scripting method restart void
```
Comment (IamPete1) [inline libraries/AP_Scripting/generator/description/bindings.desc:361]: maybe we should make it clear that its all scripts. 
```suggestion
singleton AP_Scripting method restart_all void
```

File: libraries/AP_Scripting/AP_Scripting.cpp (hunk 10 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/AP_Scripting.cpp]: done

File: libraries/AP_Scripting/generator/description/bindings.desc (hunk 11 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/generator/description/bindings.desc]: done

**Pull request merged**